src/app/lifecycle.ts

```diff
@@ -30,48 +31,97 @@ export class AppStore {
   private readonly appStore = new Map<string, FirebaseApp>();
 
   public initializeApp(options?: AppOptions, appName: string = DEFAULT_APP_NAME): App {
+    validateAppNameFormat(appName);
+
+    let autoInit = false;
     if (typeof options === 'undefined') {
+      autoInit = true
       options = loadOptionsFromEnvVar();
       options.credential = getApplicationDefault();
     }
 
-    if (typeof appName !== 'string' || appName === '') {
-      throw new FirebaseAppError(
-        AppErrorCodes.INVALID_APP_NAME,
-        `Invalid Firebase app name "${appName}" provided. App name must be a non-empty string.`,
-      );
-    } else if (this.appStore.has(appName)) {
-      if (appName === DEFAULT_APP_NAME) {
+    // Check if an App already exists and, if so, ensure its AppOptions match
+    // those of this `initializeApp` request. 
+    if (this.appStore.has(appName)) {
+      const currentApp = this.appStore.get(appName)!;
+      // Ensure the autoInit state matches the existing app's. If not, throw.
+      if (currentApp.autoInit() !== autoInit) {
         throw new FirebaseAppError(
-          AppErrorCodes.DUPLICATE_APP,
-          'The default Firebase app already exists. This means you called initializeApp() ' +
-          'more than once without providing an app name as the second argument. In most cases ' +
-          'you only need to call initializeApp() once. But if you do want to initialize ' +
-          'multiple apps, pass a second argument to initializeApp() to give each app a unique ' +
-          'name.',
-        );
+          AppErrorCodes.INVALID_APP_OPTIONS,
+          `Firebase app named "${appName}" attempted mismatch between custom AppOptions` +
+          ' and an App created via Auto Init.'
+        )
+      } else if (autoInit) {
+        // Auto-initialization is triggered when no options were passed to
+        // initializeApp. With no options to compare, simply return the App.
+        return currentApp;
       } else {
```

Comment:
Can we simplify this by using early exits/guard clauses (and removing the nested if else statements)?
If it doesn't work in this case feel free to ignore. For example:
```js
if (this.appStore.has(appName)) {
  const currentApp = this.appStore.get(appName)!;

  if (currentApp.autoInit() !== autoInit) {
    throw new FirebaseAppError(
      AppErrorCodes.INVALID_APP_OPTIONS,
      `Firebase app named "${appName}" attempted mismatch between custom AppOptions` +
      ' and an App created via Auto Init.'
    )
  }

  if (autoInit) {
    return currentApp;
  }

if (typeof options.httpAgent !== 'undefined') {
.....
```